torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -660,10 +680,66 @@ def _run_param_init_fn():
                 check_fn = lambda k: not isinstance(k, FullyShardedDataParallel)  # noqa: E731
                 deferred_init.materialize_module(module, check_fn=check_fn)
 
+        # Check that module was placed onto a single device.
+        module_devices = set(
+            p.device for p in module.parameters() if p not in ignored_params and not isinstance(p, FlatParameter)
+        )
+
+        if len(module_devices) > 1:
+            raise RuntimeError(
+                f"FSDP only supports single device modules, but got params on {module_devices}"
+            )
+
+        # Move module to device specified. Note that this is done prior to
+        # setting compute_device to ensure that they align.
+        needs_move_back_to_cpu = False
+        if self.device_id is not None:
+            param = None
+            try:
+                param = next(module.parameters())
+                if param.device == torch.device("cpu"):
+                    module = module.to(self.device_id)
+            except StopIteration:
+                # this FSDP instance manages no parameters.
+                pass
+
+            # For GPU modules, module device should match device_id.
+            if param is not None and param.device != self.device_id:
+                raise RuntimeError(
+                    f"Module on rank {self.rank} is given device_id argument "
+                    f"{self.device_id}, but is on {param.device}. "
+                    " Either move module before FSDP init or omit device_id argument."
+                )
+        else:
+            # device_id argument is not specified
+            # If module is on CPU, move it to current device and log a warning.
+            try:
+                param = next(module.parameters())
+                if param.device == torch.device("cpu"):
+                    warnings.warn(
+                        f"Module is input on CPU, we are moving it to {torch.cuda.current_device()}"
+                        " to perform parameter verification, flattening, sharding, and will"
+                        " move it back after."
+                    )
+                    needs_move_back_to_cpu = True
+                    # Note that this will match self.compute_device, because
+                    # compute_device is computed as current_device for CPU
+                    # modules.
+                    module = module.to(torch.cuda.current_device())
+            except StopIteration:
+                # this FSDP instance manages no parameters
+                needs_move_back_to_cpu = False
+
+
         # device for computation, if module is on GPU, use module.device;
         # if module is on CPU, use current device;
         self.compute_device = _get_default_cuda_device(module)
 
+        # if device_id is specified, ensure it is the same
+        assert (
```

Comment:
Curious: I am guessing that this `assert` is meant as an internal `assert` and that you updated the `return` of `_get_default_cuda_device()` so to avoid the case where `torch.device("cuda") != torch.device("cuda:0")` even though they will be redirected to the same logical device?

Reply:
My concern is what if the _user_ passes in `torch.device("cuda")`? It correctly has type `Union[int, torch.device]`, so it follows the API. However, would this `assert` be triggered given the new `_get_default_cuda_device()` `return`? (E.g. if `self.compute_device` is `torch.device("cuda:0")`, but `self.device_id` is `torch.device("cuda")`.)

More broadly as well, does this API require the user to call `torch.cuda.set_device()`? There has been confusion about that method's usage (https://github.com/pytorch/pytorch/issues/50112), so if we recommend it as part of the FSDP workflow, we may want to clarify that somewhere.

Reply:
The following is a bit unfortunate, though maybe I'm missing some context:

```
>>> import torch
>>> torch.cuda.set_device(0)
>>> z = torch.device("cuda")
>>> m = torch.device("cuda", torch.cuda.current_device())
>>> z == m
False
```

I think as guidance we can:

1) Encourage user to call torch.cuda.set_device() in general for FSDP / DDP training
2) if user passes in torch.device("cuda"), set `device_id` to `torch.device("cuda", torch.cuda.current_device())`, to ensure that the index is set?

Reply:
> The following is a bit unfortunate, though maybe I'm missing some context:

I am not sure if there is a good way to address this. I encountered similar issues in the past and was unable to come up with anything.

Nonetheless, the guidance you outlined sounds great to me!